File: src/improvement-evidence.ts
import type {
  AutonomyPolicy,
  ChangeSchema,
  Comparison,
  DatasetRow,
  DatasetSnapshot,
  MetricGate,
} from "./improvement-contracts.js";
import type { JsonValue } from "./contracts.js";
import {
  enumeration,
  fail,
  finite,
  hash,
  integer,
  json,
  nonempty,
  object,
  timestamp,
} from "./utils.js";
import { fields, target } from "./validation.js";

export function validateSchema(schema: ChangeSchema): void {
  fields(schema, [
    "type",
    "properties",
    "required",
    "additionalProperties",
    "enum",
    "minimum",
    "maximum",
    "maxLength",
  ]);
  enumeration(
    schema.type,
    ["object", "string", "number", "boolean"],
    "schema type",
  );
  if (
    schema.enum !== undefined &&
    (!Array.isArray(schema.enum) || !schema.enum.length)
  )
    fail("invalid_policy", "Empty enum.");
  if (schema.type === "object") {
    object(schema.properties, "schema properties");
    if (
      schema.additionalProperties !== false ||
      !Array.isArray(schema.required)
    )
      fail(
        "invalid_policy",
        "Objects require explicit required fields and additionalProperties:false.",
      );
    for (const name of schema.required)
      if (!Object.hasOwn(schema.properties!, name))
        fail("invalid_policy", "Unknown required field.");
    for (const value of Object.values(schema.properties!))
      validateSchema(value);
  }
  if (schema.minimum !== undefined) finite(schema.minimum, "minimum");
  if (schema.maximum !== undefined) finite(schema.maximum, "maximum");
  if (
    schema.minimum !== undefined &&
    schema.maximum !== undefined &&
    schema.minimum > schema.maximum
  )
    fail("invalid_policy", "Reversed bounds.");
  if (schema.maxLength !== undefined) integer(schema.maxLength, "maxLength");
  if (
    schema.type === "string" &&
    !schema.enum &&
    schema.maxLength === undefined
  )
    fail(
      "invalid_policy",
      "Prompt fragments require an explicit length bound.",
    );
  if (
    schema.type === "number" &&
    !schema.enum &&
    (schema.minimum === undefined || schema.maximum === undefined)
  )
    fail("invalid_policy", "Numeric changes require explicit bounds.");
}
export function validateChange(value: JsonValue, schema: ChangeSchema): void {
  json(value);
  if (schema.enum && !schema.enum.some((v) => hash(v) === hash(value)))
    fail("forbidden_change", "Value outside enum.");
  if (schema.type === "object") {
    object(value, "change");
    for (const name of schema.required!)
      if (!Object.hasOwn(value, name))
        fail("forbidden_change", `Missing ${name}.`);
    for (const [key, v] of Object.entries(value)) {
      if (!Object.hasOwn(schema.properties!, key))
        fail("forbidden_change", `Field ${key} is not permitted.`);
      validateChange(v as JsonValue, schema.properties![key]!);
    }
  } else if (typeof value !== schema.type)
    fail("forbidden_change", "Change type mismatch.");
  if (typeof value === "number") {
    finite(value, "change");
    if (
      (schema.minimum !== undefined && value < schema.minimum) ||
      (schema.maximum !== undefined && value > schema.maximum)
    )
      fail("forbidden_change", "Change outside bounds.");
  }
  if (
    typeof value === "string" &&
    schema.maxLength !== undefined &&
    Array.from(value).length > schema.maxLength
  )
    fail("forbidden_change", "String limit exceeded.");
}
function gates(g: MetricGate[]): void {
  if (!Array.isArray(g) || !g.length)
    fail("invalid_policy", "Independent guardrails required.");
  for (const item of g) {
    fields(item, ["metric", "comparator", "value"]);
    nonempty(item.metric, "metric");
    enumeration(item.comparator, ["gte", "lte"], "comparator");
    finite(item.value, "gate");
  }
}
export function validatePolicy(p: AutonomyPolicy): void {
  json(p);
  fields(p, [
    "version",
    "target",
    "changeSchema",
    "objective",
    "guardrails",
    "trustedSources",
    "minimumNewUnits",
    "minimumValidationUnits",
    "minimumAuditUnits",
    "outcomeMaturityMs",
    "maximumRows",
    "daily",
    "callbackReservation",
    "cooldownMs",
    "rollout",
    "observation",
    "onInsufficientEvidence",
    "onUncertainDeployment",
    "maximumCandidates",
    "maximumProposalCalls",
    "maximumConsecutiveFailures",
    "auditMethod",
  ]);
  nonempty(p.version, "policy version");
  target(p.target);
  enumeration(p.target.kind, ["prompt", "routing"], "autonomous target");
  validateSchema(p.changeSchema);
  fields(p.objective, [
    "metric",
    "direction",
    "minimumImprovement",
    "range",
    "alpha",
  ]);
  nonempty(p.objective.metric, "objective");
  enumeration(p.objective.direction, ["maximize", "minimize"], "direction");
  finite(p.objective.minimumImprovement, "minimumImprovement", 0);
  if (!Array.isArray(p.objective.range) || p.objective.range.length !== 2)
    fail("invalid_policy", "Metric range required.");
  p.objective.range.forEach((v) => finite(v, "range"));
  if (p.objective.range[1] <= p.objective.range[0])
    fail("invalid_policy", "Range must increase.");
  finite(p.objective.alpha, "alpha");
  if (p.objective.alpha <= 0 || p.objective.alpha >= 1)
    fail("invalid_policy", "alpha must be between 0 and 1.");
  gates(p.guardrails);
  gates(p.observation?.guardrails);
  if (!Array.isArray(p.trustedSources) || !p.trustedSources.length)
    fail("invalid_policy", "Trusted feedback sources required.");
  p.trustedSources.forEach((s) => nonempty(s, "source"));
  for (const key of [
    "minimumNewUnits",
    "minimumValidationUnits",
    "minimumAuditUnits",
    "maximumRows",
  ] as const)
    integer(p[key], key);
  for (const key of ["outcomeMaturityMs", "cooldownMs"] as const)
    integer(p[key], key, 0);
  fields(p.daily, ["modelRequests", "tokens", "deployments"]);
  fields(p.callbackReservation, ["modelRequests", "tokens"]);
  for (const key of ["modelRequests", "tokens", "deployments"] as const)
    integer(p.daily[key], key, 0);
  for (const key of ["modelRequests", "tokens"] as const)
    integer(p.callbackReservation[key], key, 0);
  if (p.daily.deployments < 1)
    fail("invalid_policy", "A finite positive deployment ceiling is required.");
  fields(p.rollout, ["mode"]);
  enumeration(p.rollout.mode, ["immediate", "canary"], "rollout mode");
  fields(p.observation, [
    "minimumDurationMs",
    "timeoutMs",
    "minimumUnits",
    "guardrails",
    "onTimeout",
    "onFailure",
  ]);
  integer(p.observation.minimumDurationMs, "minimumDurationMs", 0);
  integer(p.observation.timeoutMs, "timeoutMs");
  integer(p.observation.minimumUnits, "minimumUnits");
  if (p.observation.timeoutMs <= p.observation.minimumDurationMs)
    fail("invalid_policy", "Observation timeout must exceed minimum duration.");
  if (
    p.observation.onTimeout !== "rollback" ||
    p.observation.onFailure !== "rollback" ||
    p.onInsufficientEvidence !== "wait" ||
    p.onUncertainDeployment !== "reconcile"
  )
    fail("invalid_policy", "Unsupported unsafe fallback.");
  integer(p.maximumCandidates ?? 3, "maximumCandidates");
  integer(p.maximumProposalCalls ?? 3, "maximumProposalCalls");
  integer(p.maximumConsecutiveFailures ?? 3, "maximumConsecutiveFailures");
  if (p.auditMethod) {
    fields(p.auditMethod, ["name", "version"]);
    nonempty(p.auditMethod.name, "audit method");
    nonempty(p.auditMethod.version, "audit method version");
  }
}
/** Reject leakage; quarantine conflicting corrections and deduplicate episodes before scoring. */
export function prepareDataset(
  raw: DatasetSnapshot,
  p: AutonomyPolicy,
  now: string,
): DatasetSnapshot {
  json(raw, 16 * 1024 * 1024);
  fields(raw, ["version", "optimization", "validation", "audit"]);
  nonempty(raw.version, "dataset version");
  const partitions: DatasetRow[][] = [];
  const seenIds = new Set<string>(),
    seenEpisodes = new Set<string>(),
    seenEntities = new Set<string>();
  let lastTime = -Infinity;
  for (const name of ["optimization", "validation", "audit"] as const) {
    const rows = raw[name];
    if (!Array.isArray(rows) || rows.length > p.maximumRows)
      fail("query_limit", "Invalid or oversized dataset partition.");
    const grouped = new Map<string, DatasetRow[]>(),
      localIds = new Map<string, string>();
    for (const r of rows) {
      fields(r, [
        "id",
        "episodeId",
        "entityId",
        "occurredAt",
        "observedAt",
        "source",
        "input",
        "label",
      ]);
      for (const key of ["id", "episodeId", "entityId", "source"] as const)
        nonempty(r[key], key);
      if (!Object.hasOwn(r, "input") || !Object.hasOwn(r, "label"))
        fail("invalid_input", "Missing dataset input/label.");
      timestamp(r.occurredAt, "occurredAt");
      timestamp(r.observedAt, "observedAt");
      if (Date.parse(r.observedAt) < Date.parse(r.occurredAt))
        fail("invalid_input", "Outcome predates execution.");
      if (
        seenIds.has(r.id) ||
        seenEpisodes.has(r.episodeId) ||
        seenEntities.has(r.entityId)
      )
        fail("dataset_leakage", "Dataset partitions overlap.");
      if (localIds.has(r.id) && localIds.get(r.id) !== hash(r))
        fail(
          "conflicting_correction",
          "Duplicate row ID with conflicting content.",
        );
      localIds.set(r.id, hash(r));
      if (
        !p.trustedSources.includes(r.source) ||
        Date.parse(r.observedAt) + p.outcomeMaturityMs > Date.parse(now)
      )
        continue;
      const list = grouped.get(r.episodeId) ?? [];
      list.push(r);
      grouped.set(r.episodeId, list);
    }
    const eligible: DatasetRow[] = [],
      entityUnits = new Set<string>();
    for (const episode of [...grouped.keys()].sort((a, b) =>
      Buffer.compare(Buffer.from(a), Buffer.from(b)),
    )) {
      const group = grouped.get(episode)!;
      if (new Set(group.map((r) => hash(r.label))).size !== 1) continue;
      const r = [...group].sort((a, b) =>
        Buffer.compare(Buffer.from(a.id), Buffer.from(b.id)),
      )[0]!;
      if (new Set(group.map((r) => r.entityId)).size !== 1)
        fail("invalid_input", "An episode must belong to one entity.");
      // Conservative independence: a repeated entity cannot inflate the sample count.
      if (entityUnits.has(r.entityId)) continue;
      entityUnits.add(r.entityId);
      if (Date.parse(r.occurredAt) <= lastTime)
        fail("dataset_leakage", "Partitions must be strictly time separated.");
      eligible.push(r);
    }
    for (const r of rows) {
      seenIds.add(r.id);
      seenEpisodes.add(r.episodeId);
      seenEntities.add(r.entityId);
    }
    if (eligible.length)
      lastTime = Math.max(...eligible.map((r) => Date.parse(r.occurredAt)));
    partitions.push(
      eligible.sort((a, b) =>
        Buffer.compare(Buffer.from(a.id), Buffer.from(b.id)),
      ),
    );
  }
  return {
    version: raw.version,
    optimization: partitions[0]!,
    validation: partitions[1]!,
    audit: partitions[2]!,
  };
}
export function passesGates(
  metrics: Record<string, number>,
  rules: MetricGate[],
): boolean {
  object(metrics, "metrics");
  Object.values(metrics).forEach((v) => finite(v, "metric"));
  return rules.every(
    (g) =>
      Object.hasOwn(metrics, g.metric) &&
      (g.comparator === "gte"
        ? metrics[g.metric]! >= g.value
        : metrics[g.metric]! <= g.value),
  );
}
export function compare(
  result: Comparison,
  rows: DatasetRow[],
  p: AutonomyPolicy,
  audit = false,
) {
  json(result, 16 * 1024 * 1024);
  fields(result, ["cases", "metrics", "estimatedServingCost", "uncertainty"]);
  finite(result.estimatedServingCost, "serving cost", 0);
  if (
    !Array.isArray(result.cases) ||
    result.cases.length !== rows.length ||
    new Set(result.cases.map((r) => r.id)).size !== rows.length
  )
    fail(
      "invalid_evaluation",
      "One result per independent dataset unit required.",
    );
  const ids = new Set(rows.map((r) => r.id)),
    [lo, hi] = p.objective.range;
  let a = 0,
    b = 0;
  for (const r of result.cases) {
    fields(r, ["id", "baseline", "candidate"]);
    if (!ids.has(r.id)) fail("invalid_evaluation", "Unrecognized case.");
    for (const v of [r.baseline, r.candidate]) {
      finite(v, "case", lo);
      if (v > hi) fail("invalid_evaluation", "Score outside declared range.");
    }
    a += r.baseline;
    b += r.candidate;
  }
  const n = rows.length,
    improvement = n
      ? ((b - a) / n) * (p.objective.direction === "maximize" ? 1 : -1)
      : 0;
  // Paired differences lie in [-(hi-lo), hi-lo]; fixed-sample one-sided Hoeffding bound.
  let lower = n
    ? improvement -
      (hi - lo) * Math.sqrt((2 * Math.log(1 / p.objective.alpha)) / n)
    : -Infinity;
  if (audit && p.auditMethod) {
    const u = result.uncertainty;
    fields(u, ["method", "version", "lowerBound", "assumptionsValid"]);
    finite(u.lowerBound, "custom lower bound");
    if (
      u.method !== p.auditMethod.name ||
      u.version !== p.auditMethod.version ||
      u.assumptionsValid !== true
    )
      fail(
        "invalid_evaluation",
        "Custom audit method identity/assumptions not satisfied.",
      );
    if (u.lowerBound > improvement || u.lowerBound < -(hi - lo))
      fail(
        "invalid_evaluation",
        "Custom lower bound is inconsistent with the bounded observed effect.",
      );
    lower = u.lowerBound;
  }
  const passed =
    n > 0 &&
    n >= (audit ? p.minimumAuditUnits : p.minimumValidationUnits) &&
    passesGates(result.metrics, p.guardrails) &&
    (audit ? lower : improvement) > p.objective.minimumImprovement;
  return {
    passed,
    baseline: n ? a / n : 0,
    candidate: n ? b / n : 0,
    improvement,
    lower: Number.isFinite(lower) ? lower : null,
    sampleCount: n,
  };
}

File: python/src/loopiter/improvement_evidence.py
"""Dependency-free bounded schemas, independent units and fixed-sample audit gates.

Scores are a single explicitly directed objective, not an average of unrelated metrics.
The paired Hoeffding bound assumes independent bounded units and a frozen audit set.
"""

import math
from copy import deepcopy

from . import _validation as v


def validate_schema(schema):
    v.fields(
        schema,
        [
            "type",
            "properties",
            "required",
            "additionalProperties",
            "enum",
            "minimum",
            "maximum",
            "maxLength",
        ],
    )
    v.enum(schema.get("type"), ["object", "string", "number", "boolean"], "schema type")
    if "enum" in schema and (type(schema["enum"]) is not list or not schema["enum"]):
        v.fail("invalid_policy", "Empty enum.")
    if schema["type"] == "object":
        v.obj(schema.get("properties"), "schema properties")
        if (
            schema.get("additionalProperties") is not False
            or type(schema.get("required")) is not list
        ):
            v.fail(
                "invalid_policy", "Objects require explicit fields and additionalProperties:false."
            )
        if any(key not in schema["properties"] for key in schema["required"]):
            v.fail("invalid_policy", "Unknown required field.")
        for child in schema["properties"].values():
            validate_schema(child)
    for name in ("minimum", "maximum"):
        if name in schema:
            v.finite(schema[name], name)
    if schema.get("minimum", -math.inf) > schema.get("maximum", math.inf):
        v.fail("invalid_policy", "Reversed bounds.")
    if "maxLength" in schema:
        v.integer(schema["maxLength"], "maxLength")
    if schema["type"] == "string" and "enum" not in schema and "maxLength" not in schema:
        v.fail("invalid_policy", "Prompt fragments require an explicit length bound.")
    if (
        schema["type"] == "number"
        and "enum" not in schema
        and not {"minimum", "maximum"} <= set(schema)
    ):
        v.fail("invalid_policy", "Numeric changes require explicit bounds.")


def validate_change(value, schema):
    v.json_value(value)
    if "enum" in schema and not any(
        v.fingerprint(value) == v.fingerprint(x) for x in schema["enum"]
    ):
        v.fail("forbidden_change", "Value outside enum.")
    kind = schema["type"]
    if kind == "object":
        v.obj(value, "change")
        if set(schema["required"]) - set(value) or set(value) - set(schema["properties"]):
            v.fail("forbidden_change", "Unknown or missing change fields.")
        for name, child in value.items():
            validate_change(child, schema["properties"][name])
    elif kind == "number":
        v.finite(value, "change")
        if not schema.get("minimum", -math.inf) <= value <= schema.get("maximum", math.inf):
            v.fail("forbidden_change", "Change outside bounds.")
    elif type(value) is not {"string": str, "boolean": bool}[kind]:
        v.fail("forbidden_change", "Change type mismatch.")
    if type(value) is str and len(value) > schema.get("maxLength", math.inf):
        v.fail("forbidden_change", "String limit exceeded.")


def _gates(rules):
    if type(rules) is not list or not rules:
        v.fail("invalid_policy", "Independent guardrails required.")
    for gate in rules:
        v.fields(gate, ["metric", "comparator", "value"], ("metric", "comparator", "value"))
        v.nonempty(gate["metric"], "metric")
        v.enum(gate["comparator"], ["gte", "lte"], "comparator")
        v.finite(gate["value"], "gate")


def validate_policy(p):
    v.json_value(p)
    names = (
        "version target change_schema objective guardrails trusted_sources minimum_new_units "
        "minimum_validation_units minimum_audit_units outcome_maturity_ms maximum_rows daily "
        "callback_reservation cooldown_ms rollout observation on_insufficient_evidence "
        "on_uncertain_deployment"
    ).split()
    v.fields(
        p,
        names
        + [
            "maximum_candidates",
            "maximum_proposal_calls",
            "maximum_consecutive_failures",
            "audit_method",
        ],
        tuple(names),
    )
    v.nonempty(p["version"], "policy version")
    v.target(p["target"])
    v.enum(p["target"]["kind"], ["prompt", "routing"], "autonomous target")
    validate_schema(p["change_schema"])
    o = p["objective"]
    v.fields(
        o,
        ["metric", "direction", "minimum_improvement", "range", "alpha"],
        ("metric", "direction", "minimum_improvement", "range", "alpha"),
    )
    v.nonempty(o["metric"], "objective")
    v.enum(o["direction"], ["maximize", "minimize"], "direction")
    v.finite(o["minimum_improvement"], "minimum improvement", 0)
    if type(o["range"]) is not list or len(o["range"]) != 2:
        v.fail("invalid_policy", "Metric range required.")
    for x in o["range"]:
        v.finite(x, "range")
    v.finite(o["alpha"], "alpha")
    if o["range"][1] <= o["range"][0] or not 0 < o["alpha"] < 1:
        v.fail("invalid_policy", "Invalid range/alpha.")
    _gates(p["guardrails"])
    if type(p["trusted_sources"]) is not list or not p["trusted_sources"]:
        v.fail("invalid_policy", "Trusted sources required.")
    for source in p["trusted_sources"]:
        v.nonempty(source, "source")
    for key in (
        "minimum_new_units",
        "minimum_validation_units",
        "minimum_audit_units",
        "maximum_rows",
    ):
        v.integer(p[key], key)
    for key in ("outcome_maturity_ms", "cooldown_ms"):
        v.integer(p[key], key, 0)
    for name, keys in [
        ("daily", ["model_requests", "tokens", "deployments"]),
        ("callback_reservation", ["model_requests", "tokens"]),
    ]:
        v.fields(p[name], keys, tuple(keys))
        for key in keys:
            v.integer(p[name][key], key, 0)
    if p["daily"]["deployments"] < 1:
        v.fail("invalid_policy", "A positive deployment ceiling is required.")
    v.fields(p["rollout"], ["mode"], ("mode",))
    v.enum(p["rollout"]["mode"], ["immediate", "canary"], "rollout mode")
    obs = p["observation"]
    v.fields(
        obs,
        [
            "minimum_duration_ms",
            "timeout_ms",
            "minimum_units",
            "guardrails",
            "on_timeout",
            "on_failure",
        ],
        (
            "minimum_duration_ms",
            "timeout_ms",
            "minimum_units",
            "guardrails",
            "on_timeout",
            "on_failure",
        ),
    )
    v.integer(obs["minimum_duration_ms"], "minimum duration", 0)
    v.integer(obs["timeout_ms"], "timeout")
    v.integer(obs["minimum_units"], "minimum units")
    _gates(obs["guardrails"])
    if obs["timeout_ms"] <= obs["minimum_duration_ms"]:
        v.fail("invalid_policy", "Observation timeout must exceed minimum duration.")
    if (
        obs["on_timeout"],
        obs["on_failure"],
        p["on_insufficient_evidence"],
        p["on_uncertain_deployment"],
    ) != ("rollback", "rollback", "wait", "reconcile"):
        v.fail("invalid_policy", "Unsupported unsafe fallback.")
    for key in ("maximum_candidates", "maximum_proposal_calls", "maximum_consecutive_failures"):
        v.integer(p.get(key, 3), key)
    if "audit_method" in p:
        v.fields(p["audit_method"], ["name", "version"], ("name", "version"))
        v.nonempty(p["audit_method"]["name"], "audit method")
        v.nonempty(p["audit_method"]["version"], "audit method version")


def prepare_dataset(raw, policy, now):
    v.json_value(raw, 16 * 1024 * 1024)
    v.fields(
        raw,
        ["version", "optimization", "validation", "audit"],
        ("version", "optimization", "validation", "audit"),
    )
    v.nonempty(raw["version"], "dataset version")
    result = {"version": raw["version"]}
    ids, episodes, entities = set(), set(), set()
    last_time = -math.inf
    now_ms = v.timestamp(now).timestamp() * 1000
    for name in ("optimization", "validation", "audit"):
        rows = raw[name]
        if type(rows) is not list or len(rows) > policy["maximum_rows"]:
            v.fail("query_limit", "Invalid or oversized dataset partition.")
        grouped, local_ids = {}, {}
        for row in rows:
            names = "id episode_id entity_id occurred_at observed_at source input label".split()
            v.fields(row, names, tuple(names))
            for key in ("id", "episode_id", "entity_id", "source"):
                v.nonempty(row[key], key)
            occurred = v.timestamp(row["occurred_at"]).timestamp() * 1000
            observed = v.timestamp(row["observed_at"]).timestamp() * 1000
            if observed < occurred:
                v.fail("invalid_input", "Outcome predates execution.")
            if row["id"] in ids or row["episode_id"] in episodes or row["entity_id"] in entities:
                v.fail("dataset_leakage", "Dataset partitions overlap.")
            digest = v.fingerprint(row)
            if row["id"] in local_ids and local_ids[row["id"]] != digest:
                v.fail("conflicting_correction", "Duplicate row ID with conflicting content.")
            local_ids[row["id"]] = digest
            if (
                row["source"] not in policy["trusted_sources"]
                or observed + policy["outcome_maturity_ms"] > now_ms
            ):
                continue
            grouped.setdefault(row["episode_id"], []).append(row)
        eligible, entity_units = [], set()
        for episode in sorted(grouped):
            group = grouped[episode]
            if len({v.fingerprint(row["label"]) for row in group}) != 1:
                continue
            row = sorted(group, key=lambda r: r["id"])[0]
            if len({r["entity_id"] for r in group}) != 1:
                v.fail("invalid_input", "An episode must belong to one entity.")
            if row["entity_id"] in entity_units:
                continue
            entity_units.add(row["entity_id"])
            if v.timestamp(row["occurred_at"]).timestamp() <= last_time:
                v.fail("dataset_leakage", "Partitions must be strictly time separated.")
            eligible.append(row)
        ids.update(row["id"] for row in rows)
        episodes.update(row["episode_id"] for row in rows)
        entities.update(row["entity_id"] for row in rows)
        if eligible:
            last_time = max(v.timestamp(row["occurred_at"]).timestamp() for row in eligible)
        result[name] = deepcopy(sorted(eligible, key=lambda r: r["id"]))
    return result


def passes_gates(metrics, rules):
    v.obj(metrics, "metrics")
    for value in metrics.values():
        v.finite(value, "metric")
    return all(
        g["metric"] in metrics
        and (
            metrics[g["metric"]] >= g["value"]
            if g["comparator"] == "gte"
            else metrics[g["metric"]] <= g["value"]
        )
        for g in rules
    )


def compare(result, rows, p, audit=False):
    v.json_value(result, 16 * 1024 * 1024)
    v.fields(
        result,
        ["cases", "metrics", "estimated_serving_cost", "uncertainty"],
        ("cases", "metrics", "estimated_serving_cost"),
    )
    v.finite(result["estimated_serving_cost"], "serving cost", 0)
    cases = result["cases"]
    if type(cases) is not list or len(cases) != len(rows):
        v.fail("invalid_evaluation", "One result per independent dataset unit required.")
    ids, seen, a, b = {r["id"] for r in rows}, set(), 0, 0
    lo, hi = p["objective"]["range"]
    for row in cases:
        v.fields(row, ["id", "baseline", "candidate"], ("id", "baseline", "candidate"))
        if row["id"] not in ids or row["id"] in seen:
            v.fail("invalid_evaluation", "Unrecognized/duplicate case.")
        seen.add(row["id"])
        for value in (row["baseline"], row["candidate"]):
            v.finite(value, "case", lo)
            if value > hi:
                v.fail("invalid_evaluation", "Score outside declared range.")
        a += row["baseline"]
        b += row["candidate"]
    n = len(rows)
    improvement = (b - a) / n * (1 if p["objective"]["direction"] == "maximize" else -1) if n else 0
    lower = (
        improvement - (hi - lo) * math.sqrt(2 * math.log(1 / p["objective"]["alpha"]) / n)
        if n
        else None
    )
    if audit and "audit_method" in p:
        u = result.get("uncertainty")
        v.fields(
            u,
            ["method", "version", "lower_bound", "assumptions_valid"],
            ("method", "version", "lower_bound", "assumptions_valid"),
        )
        v.finite(u["lower_bound"], "custom lower bound")
        if (
            u["method"] != p["audit_method"]["name"]
            or u["version"] != p["audit_method"]["version"]
            or u["assumptions_valid"] is not True
        ):
            v.fail("invalid_evaluation", "Custom audit identity/assumptions not satisfied.")
        if not -(hi - lo) <= u["lower_bound"] <= improvement:
            v.fail("invalid_evaluation", "Custom lower bound inconsistent with observed effect.")
        lower = u["lower_bound"]
    passed = bool(
        n
        and n >= p["minimum_audit_units" if audit else "minimum_validation_units"]
        and passes_gates(result["metrics"], p["guardrails"])
        and (lower if audit else improvement) > p["objective"]["minimum_improvement"]
    )
    return {
        "passed": passed,
        "baseline": a / n if n else 0,
        "candidate": b / n if n else 0,
        "improvement": improvement,
        "lower": lower,
        "sample_count": n,
    }
